### PR TITLE
cli: add global cache for binaries

### LINKED ISSUE #1188: Docs site and binary install from yarn can be very slow from Korea
- Operating System: OSX
- Cypress Version: Any

### Current behavior:
Sometimes cypress can take up to 30 minutes to download

### How to reproduce:
Use a blazing fast internet connection in Korea

#### Docs site after a few minutes (still no fonts)
screen shot 2018-01-12 at 2 35 49 pm

#### Download

```
[21:47:48]  Downloading Cypress      21% 4487s [title changed]\n[21:47:48]  Downloading Cypress      21% 4489s [title changed]\n[21:47:49]  Downloading Cypress      21% 4491s [title changed]\n[21:47:50]  Downloading Cypress      21% 4495s [title changed]\n[21:47:51]  Downloading Cypress      21% 4498s [title changed]\n[21:47:51]  Downloading Cypress      21% 4500s [title changed]\n[21:47:51]  Downloaded Cypress      [title changed]\n[21:47:51]  Downloaded Cypress      [completed]\n[21:47:51]  Unzipping Cypress       [started]\n[21:47:51]  Unzipping Cypress       [failed]\n[21:47:51] → The Cypress App could not be unzipped.\n\nSearch for an existing issue or open a GitHub issue at\n\nhttps://github.com/cypress-io/cypress/issues\n----------\n\nError: end of central directory record signature not found\n----------\n\nPlatform: darwin (15.6.0)\nCypress Version: 1.4.1\nThe Cypress App could not be unzipped
```

The cypress docs site takes minutes to load docs sometimes.  This isn't 100% consistent, but it seems to be becoming a more frequent issue.  While I can generally use some of the docs during the loading, if I have to re-`yarn` sometimes it stops me in my tracks as a download can take 30 minutes and fail to unzip.  I can't say that this is always the case, but it is frequent and it is, as far as I can tell, unique to the cypress.io server, as I can't think of another site I experience this with.  I've also had this experience at home as well as various coffee shop wifi.

### PR DESCRIPTION
**Pull request by maxime1992**
Global cache will speed up local development when running `npm i` or `yarn`.
It'll downloaded Cypress binaries once and if available, next time will only get the cached one.

I believe it might also be safer for projects using Cypress to cache only that global cache folder + yarn or npm cache on CI instead of the whole node_modules.

This closes #1188

### CODE REVIEW

**Comment by maxime1992** [comment]
So, I think it's good now, just waiting for comments :smile:!

One thing tho: I don't know how to test the functionality **for real** (which is not ideal of course).

**Comment by brian-mann** [comment]
pinging @bahmutov 

My thoughts:

There are some existing node modules that handle the "global cache" like this - which prevents you from having to build up the `homeDir/.cypress/` pathing. We've had so many issues with Windows that we internally always look for modules that have already solved this in a cross platform way. Although this is minor - you shouldn't ever build up path slashes like you're doing because of Windows - always go through `path.join(...)`

Alternatively - Cypress itself uses the "proper" application cache of the OS which we utilize here https://github.com/cypress-io/cypress/blob/develop/packages/server/lib/util/app_data.coffee. It may make more sense to store the binaries there since we also give you a way to "View App Data" from the "File" menu from the GUI itself. It would make things more consistent and that code is already cross platform ready. We'd need to extract this file into something more generic to make it available to the CLI.

As for testing this code for real - you generally just start from another project and call directly into the CLI code by invoking `cypress/cli/bin/cypress` directly. That should make it behave as if it's been NPM installed. You could also take an existing installation and just copy/paste your updated code into the NPM module.

**Comment by brian-mann** [comment]
@bahmutov thinking about this further....

Maybe instead of caching the zipped Cypress globally and then unzipping it into the `dist` folder within the NPM module -- why wouldn't we unzip it into the global cache and then just point the CLI to use the global version?

This would save you from having Cypress (which is very large) installed on multiple projects. As long as they use the same version number they'd all use the exact binary.

We'd have to likely come up with some new CLI flags to clear the cache when you want to do reinstall the binary or maybe a `--force` flag which bypasses the cache... but this seems to not only solve this issue of caching, but also fixes other issues people have asked for (that Cypress takes up too much space across multiple project installations).

**Comment by paulfalgout** [comment]
While I like where this is headed and it will help the issue, it won't resolve #1188 which seems to be that cloudflare is abysmal in the eastern hemisphere.  Not only does re-downloading Cypress take 30 minutes, but the docs take an enormous amount of time to load.

**Comment by bahmutov** [comment]
I'm all for unzipping into same folder by cypress version. Also finding a module that can help if possible because cross platform is a nightmare.

File: cli/lib/tasks/download.js [hunk 1]
```diff
@@ -120,15 +121,21 @@ const downloadFromUrl = (options) => {
...
       debug('downloading finished')
 
-      resolve({
-        filename: options.downloadDestination,
-        downloaded: true,
-      })
+      // ... and also add it to the global cache
+      fs.copy(
+        options.downloadDestination,
+        `${home}/.cypress/${options.version}`
```
Comment (bahmutov) [inline cli/lib/tasks/download.js]: we should never concatenate paths ourselves (Windows ...)

**Comment by bahmutov** [comment]
BTW last comment: I usually do not struggle with Cypress installs because I use https://github.com/bahmutov/have-it to pull an alias from some other project that already has it installed
```
$ have -D cypress
```

**Comment by bahmutov** [comment]
@maxime1992 I will close this request in favor of https://github.com/cypress-io/cypress/issues/1300 - reusing same binary from different projects. This will avoid custom code and will do the same caching as other projects are doing (like Electron).

**Pull request closed without merging**